src/main/kotlin/frc/robot/subsystems/SwerveSubsystem.kt

```diff
@@ -0,0 +1,176 @@
+        val swerveModuleStates = if (fieldRelative) {
+            DrivetrainConstants.driveKinematics.toSwerveModuleStates(
+                ChassisSpeeds.fromFieldRelativeSpeeds(
+                    forwardMetersPerSecond,
+                    sidewaysMetersPerSecond,
+                    radiansPerSecond,
+                    Rotation2d.fromRadians(heading)))
+        } else {
+            DrivetrainConstants.driveKinematics.toSwerveModuleStates(
+                ChassisSpeeds(
+                    forwardMetersPerSecond,
+                    sidewaysMetersPerSecond,
+                    radiansPerSecond))
+        }
+
+
+
+        SwerveDriveKinematics.desaturateWheelSpeeds(swerveModuleStates, DrivetrainConstants.maxSpeedMetersPerSecond)
+
+        frontLeft.setDesiredState(swerveModuleStates[0])
+        frontRight.setDesiredState(swerveModuleStates[1])
+        rearLeft.setDesiredState(swerveModuleStates[2])
+        rearRight.setDesiredState(swerveModuleStates[3])
+    }
+
+    fun setX() {
+        frontLeft.setDesiredState(SwerveModuleState(0.0, Rotation2d.fromDegrees(-45.0)))
+        frontRight.setDesiredState(SwerveModuleState(0.0, Rotation2d.fromDegrees(45.0)))
+        rearLeft.setDesiredState(SwerveModuleState(0.0, Rotation2d.fromDegrees(45.0)))
+        rearRight.setDesiredState(SwerveModuleState(0.0, Rotation2d.fromDegrees(-45.0)))
+    }
+
+    fun setZero() {
+        frontLeft.setDesiredState(SwerveModuleState(0.0, Rotation2d.fromDegrees(0.0)))
+        frontRight.setDesiredState(SwerveModuleState(0.0, Rotation2d.fromDegrees(0.0)))
+        rearLeft.setDesiredState(SwerveModuleState(0.0, Rotation2d.fromDegrees(0.0)))
+        rearRight.setDesiredState(SwerveModuleState(0.0, Rotation2d.fromDegrees(0.0)))
+    }
+
+    fun zeroGyro() {
+        gyro.reset()
+    }
+
+    fun setModuleStates(desiredStates: Array<SwerveModuleState>) {
+        SwerveDriveKinematics.desaturateWheelSpeeds(desiredStates, DrivetrainConstants.maxSpeedMetersPerSecond)
+
+        frontLeft.setDesiredState(desiredStates[0])
+        frontRight.setDesiredState(desiredStates[1])
+        rearLeft.setDesiredState(desiredStates[2])
+        rearRight.setDesiredState(desiredStates[3])
+    }
+
+    fun resetEncoders() {
+        frontLeft.resetEncoders()
+        frontRight.resetEncoders()
+        rearLeft.resetEncoders()
+        rearRight.resetEncoders()
+    }
+
+    fun zeroHeading() {
+        gyro.reset()
+    }
+
+//    fun getHeading(): Double {
+//        return Rotation2d.fromDegrees(heading).degrees
+//    }
+
+    fun turnRate(): Double {
+        val coefficient = if (DrivetrainConstants.gyroReversed) {
+            -1.0
+        } else {
+            1.0
+        }
+        return gyro.rate * coefficient
+    }
+    fun directionDrive(swerveStates: SwerveModuleState) {
+        frontLeft.setDesiredState(swerveStates)
```

Comment:
You don't need this, this is what setModulesState does

Reply:
Done